bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/commands/CommandImagePersistence.java

```diff
@@ -126,6 +132,34 @@ private static void readImagesFromRegistry(final IConfigurationElement[] configu
 				"Warnings while parsing the images from the 'org.eclipse.ui.commandImages' extension point."); //$NON-NLS-1$
 	}
 
+	private static void readImagesFromApplicationModel(final CommandImageManager commandImageManager) {
+
+		MApplication application = E4Workbench.getServiceContext().get(MApplication.class);
+
+		for (MCommand command : application.getCommands()) {
+
+			if (command.getPersistedState().containsKey(E4_COMMAND_IMAGE_KEY)) {
```

Comment:
It seems other parts of the code using an dedicated `iconUri` (with a dialog to search / select icons) is the persisted state really the right place to do this?
I suspect that this way it might be possible to set an icon but hardly anyone will find / use that?!

Reply:
The dedicated `iconUri ` field would require a change to the Application Model, and to be honest I don't know how to do this and if there is a process to discuss such a change, 
The Persisted State is maybe not the best place to put it is already used for some special features like this.

Reply:
> The dedicated iconUri field would require a change to the Application Model, and to be honest I don't know how to do this and if there is a process to discuss such a change,

@vogella or @merks can help with the model change, I think this is the right way to solve this, the persisted state seems not an appropriate choice here:

1. not visible to users so most likely no one will ever discover your new feature
2. no user assistance in available fields
3. no validation
4. not consistent across other usages for a similar feature (menus, parts, ...) where icons are provided with E4

Reply:
I agree with you that it is not the right place, i will have a look into the model change.

Reply:
I can help with changing the model, but I did not design the model, so I don't know the proper design for this. It would seem like three new attributes are needed, **but** I don't see anything similar to that in the model already.  E.g., I don't know how toolbar items get their icon from the model in a pure e4 application.  How does that work currently?

Reply:
Currently, the icon for a toolbar item or Part is set as an attribute of MUILabel from which the Part or ToolItem inherits. 
As MUILabel provides more functionality than needed, I tried to add an attribute directly to MCommand. Which looks fine,

The only issue I have now is that the code generation generates a newer style than the current code, which creates more changes than necessary. 
Should I commit the whole changes, or strip it down to only the relevant ones?

Reply:
Yes, that's the only related thing that I saw, but there I did not see other things one might expect from your changes, e.g., a disabled icon or a hover icon.  That that makes me wonder why are those needed for a command but not needed elsewhere in the UI model where I would assume the same kinds of needs arise.

Reply:
They are usually not needed, there is an option to set those in the transient data of an Element (that is really well hidden).
That is optional and only needed when the generation of the disabled and hover icon should be overridden.

I added them for feature completeness to e3. But for the solution with the application model, I would keep them in the persisted state and move only the main icon to an attribute and add the option to the E4 Model Editor.

Reply:
I see!  

So if you change the model to add a new attribute, you'd normally increment the minor version, but that's already been done for the 4.28 release cycle.   You'd add something that looks like this in the UIElement.ecore:
```
      <eStructuralFeatures xsi:type="ecore:EAttribute" name="iconURI" eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EString">
        <eAnnotations source="http://www.eclipse.org/emf/2002/GenModel">
          <details key="documentation" value="&lt;p>&#xD;&#xA;This field contains a fully qualified URL defining the path to an Image to display&#xD;&#xA;for this element.&#xD;&#xA;&lt;/p>&#xD;&#xA;@since 2.4.0"/>
        </eAnnotations>
      </eStructuralFeatures>
```
Note the @since 2.4.0 which will be generated into all the parts of the code that change.  When you open the *.genmodel, it will be dirty and will need to be saved.  You can generate the code from there:

![image](https://user-images.githubusercontent.com/208716/227868556-122d349f-fa3c-4350-a46c-83817de1d781.png)

Of course you have to have EMF's generator installed, but the Oomph setup includes that already.

If you need other help, just ask.

Reply:
@merks @laeubi I updated the E4 Model with an attribute for the iconUrl and extended the Model Editor to edit the field.

Reply:
The model changes look fine now.